File: mobile/ios/Actnet/Sources/Views/Chats/GroupDetailView.swift
import SwiftUI

/// Group detail screen (docs/03-groups.md): member list, admin role changes,
/// disappearing-message timer, and leave-group. Admin-only controls are shown
/// only when the current user is an admin; the server enforces the same.
struct GroupDetailView: View {
    @EnvironmentObject var appState: AppState
    @Environment(\.dismiss) private var dismiss

    let groupId: String
    let accountId: String

    @State private var summary: GroupSummaryFfi?
    @State private var loading: Bool = true
    @State private var errorMessage: String?
    /// Bound to the timer picker; seeded from the loaded state so the initial
    /// assignment doesn't fire a spurious change (we guard on the loaded value).
    @State private var expirySeconds: UInt32 = 0
    @State private var showRename = false
    @State private var renameText = ""

    /// True when the current user is an admin of this group — gates the
    /// role/timer controls.
    private var amAdmin: Bool {
        summary?.members.first(where: { $0.did == accountId })?.role == 1
    }

    /// Whether we're still a member — false once we've left (docs/53 §Leave).
    /// Gates the "Leave group" button; the rest of the screen renders read-only
    /// from cached state.
    private var amMember: Bool {
        summary?.members.contains(where: { $0.did == accountId }) ?? false
    }

    var body: some View {
        Group {
            if let s = summary {
                List {
                    Section {
                        VStack(spacing: 12) {
                            groupAvatarView(s)
                                .frame(maxWidth: .infinity)
                            VStack(alignment: .leading, spacing: 4) {
                                HStack {
                                    Text(s.title.isEmpty ? "Group" : s.title).font(.headline)
                                    if amAdmin {
                                        Spacer()
                                        Button("Rename") {
                                            renameText = s.title
                                            showRename = true
                                        }
                                        .font(.subheadline)
                                    }
                                }
                                if !s.description.isEmpty {
                                    Text(s.description).font(.subheadline).foregroundStyle(.secondary)
                                }
                                Text("Revision \(s.revision)").font(.caption).foregroundStyle(.secondary)
                            }
                        }
                    }
                    Section("Disappearing messages") {
                        if amAdmin {
                            DisappearingMessagesPicker(seconds: $expirySeconds)
                                .onChange(of: expirySeconds) { _, newValue in
                                    // Ignore the initial seeding assignment; only
                                    // act on a real user change away from the
                                    // loaded value.
                                    if newValue != s.expirySeconds { setExpiry(newValue) }
                                }
                        } else {
                            HStack {
                                Text("Timer")
                                Spacer()
                                Text(DisappearingMessagesPicker.label(for: s.expirySeconds))
                                    .foregroundStyle(.secondary)
                            }
                        }
                    }
                    Section("Members (\(s.members.count))") {
                        ForEach(orderedMembers(s.members), id: \.encryptedMemberId) { member in
                            // Tapping a member row opens a menu: message them
                            // (note-to-self for your own row), plus admin
                            // promote/demote for anyone but yourself.
                            Menu {
                                Button {
                                    openDm(member)
                                } label: {
                                    Label(dmLabel(member), systemImage: "bubble.left")
                                }
                                // Copy this member as a contact card (docs/35):
                                // writes {did, name} to the clipboard; paste it
                                // into any conversation to share them. Use the
                                // *resolved* name, not `memberName` — for your own
                                // row that would be "You"; sharing your own card
                                // must carry your real display name.
                                Button {
                                    ContactPasteboard.write(
                                        did: member.did,
                                        name: appState.resolvedName(for: member.did, accountId: accountId)
                                    )
                                } label: {
                                    Label("Copy contact", systemImage: "doc.on.doc")
                                }
                                if amAdmin && member.did != accountId {
                                    if member.role == 1 {
                                        Button {
                                            changeRole(member, toAdmin: false)
                                        } label: {
                                            Label("Remove admin", systemImage: "person.badge.minus")
                                        }
                                    } else {
                                        Button {
                                            changeRole(member, toAdmin: true)
                                        } label: {
                                            Label("Make admin", systemImage: "person.badge.shield.checkmark")
                                        }
                                    }
                                }
                            } label: {
                                HStack(spacing: 10) {
                                    ContactAvatar(
                                        name: memberName(member),
                                        imageData: appState.avatar(for: member.did, accountId: accountId),
                                        isBot: isBot(member),
                                        size: 32
                                    )
                                    Text(memberName(member))
                                        .lineLimit(1)
                                        .foregroundStyle(.primary)
                                    Spacer()
                                    if member.role == 1 {
                                        Text("Admin").font(.caption2)
                                            .padding(.horizontal, 6).padding(.vertical, 2)
                                            .background(Color.accentColor.opacity(0.15))
                                            .clipShape(Capsule())
                                    }
                                }
                                .contentShape(Rectangle())
                            }
                            .tint(.primary)
                        }
                    }
                    if !s.pendingInvites.isEmpty {
                        Section("Pending invites (\(s.pendingInvites.count))") {
                            ForEach(s.pendingInvites, id: \.encryptedMemberId) { p in
                                Text(p.encryptedMemberId).font(.caption).lineLimit(1)
                            }
                        }
                    }
                    if amMember {
                        Section {
                            Button(role: .destructive) {
                                leaveGroup()
                            } label: {
                                Label("Leave group", systemImage: "rectangle.portrait.and.arrow.right")
                            }
                        }
                    } else {
                        Section {
                            Text("You left this group.")
                                .font(.caption)
                                .foregroundStyle(.secondary)
                        }
                    }
                }
                .scrollContentBackground(.hidden)
                .background(Color.avPaper)
            } else if loading {
                ProgressView().padding()
            } else if let error = errorMessage {
                ContentUnavailableView(
                    "Couldn't load group",
                    systemImage: "exclamationmark.triangle",
                    description: Text(error)
                )
            }
        }
        .background(Color.avPaper)
        .navigationTitle("Group info")
        .navigationBarTitleDisplayMode(.inline)
        .alert("Rename group", isPresented: $showRename) {
            TextField("Group name", text: $renameText)
            Button("Cancel", role: .cancel) {}
            Button("Save") { renameGroup(renameText) }
                .disabled(renameText.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty)
        }
        .task { await load() }
    }

    /// Rename the group (admin-only; server-enforced). app-core emits the
    /// "changed the group name to X" timeline entry, so the conversation updates.
    /// Group avatar (docs/55): editable for admins (gated by `modify_title_role`,
    /// same as rename), read-only for everyone else.
    @ViewBuilder
    private func groupAvatarView(_ s: GroupSummaryFfi) -> some View {
        let current = appState.groupAvatar(groupId: groupId, accountId: accountId)
        let name = s.title.isEmpty ? "Group" : s.title
        if amAdmin {
            EditableAvatar(
                currentImage: current,
                placeholderName: name,
                size: 72,
                onPicked: { data in Task { await setGroupAvatar(data) } },
                onRemove: current == nil ? nil : { Task { await removeGroupAvatar() } }
            )
        } else {
            ContactAvatar(name: name, imageData: current, size: 72)
        }
    }

    private func setGroupAvatar(_ data: Data) async {
        do {
            try await appState.setGroupAvatar(data, groupId: groupId, accountId: accountId)
            await load()
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    private func removeGroupAvatar() async {
        do {
            try await appState.removeGroupAvatar(groupId: groupId, accountId: accountId)
            await load()
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    private func renameGroup(_ name: String) {
        let trimmed = name.trimmingCharacters(in: .whitespacesAndNewlines)
        // Group names must not be empty, and skip a no-op rename.
        guard let core = appState.core(accountId: accountId),
              !trimmed.isEmpty,
              trimmed != (summary?.title ?? "")
        else { return }
        let gid = groupId
        Task {
            do {
                try await Task.detached {
                    try core.setGroupTitle(groupId: gid, newTitle: trimmed)
                }.value
                await load()
            } catch {
                errorMessage = error.localizedDescription
            }
        }
    }

    /// The current user sorts first; everyone else keeps server order.
    private func orderedMembers(_ members: [GroupMemberFfi]) -> [GroupMemberFfi] {
        members.sorted { a, _ in a.did == accountId }
    }

    /// Member's display name, or "You" for the current user.
    private func memberName(_ member: GroupMemberFfi) -> String {
        member.did == accountId
            ? "You"
            : appState.resolvedName(for: member.did, accountId: accountId)
    }

    /// Menu label for the "message this member" action. A DM to your own
    /// identity is a note-to-self (docs/04 §5.5), like Signal.
    private func dmLabel(_ member: GroupMemberFfi) -> String {
        member.did == accountId
            ? "Note to self"
            : "Message \(appState.resolvedName(for: member.did, accountId: accountId))"
    }

    /// Open (or create) the DM with a group member and navigate to it. Resets
    /// the chats nav stack to the DM, popping this group-info screen.
    private func openDm(_ member: GroupMemberFfi) {
        let conv = appState.findOrCreateDMConversation(recipientDid: member.did, accountId: accountId)
        appState.selectedTab = .chats
        appState.navigateToConversation = conv
    }

    /// Whether a member is a bot, for the hexagon avatar frame
    /// (docs/54-bot-presentation.md). The local user is never a bot.
    private func isBot(_ member: GroupMemberFfi) -> Bool {
        member.did != accountId && appState.isBot(member.did, accountId: accountId)
    }

    private func load() async {
        loading = true
        defer { loading = false }
        guard let core = appState.core(accountId: accountId) else {
            errorMessage = "No core for account"
            return
        }
        let gid = groupId
        do {
            let s = try await Task.detached {
                try core.fetchGroupState(groupId: gid)
            }.value
            self.summary = s
            self.expirySeconds = s.expirySeconds
        } catch {
            // The server fetch is membership-gated, so it 404s for a group we've
            // left (docs/53 §Leave). Fall back to the last-known cached state so
            // the info screen still renders, read-only, instead of erroring.
            let cached = try? await Task.detached(operation: {
                try core.cachedGroupState(groupId: gid)
            }).value
            if let cached {
                self.summary = cached
                self.expirySeconds = cached.expirySeconds
            } else {
                errorMessage = error.localizedDescription
            }
        }
    }

    /// Promote or demote a member (admin-only; server-enforced). The system
    /// timeline entry is emitted by app-core, so the conversation updates too.
    private func changeRole(_ member: GroupMemberFfi, toAdmin: Bool) {
        guard let core = appState.core(accountId: accountId) else { return }
        let gid = groupId
        let emi = member.encryptedMemberId
        let newRole: Int16 = toAdmin ? 1 : 0
        Task {
            do {
                try await Task.detached {
                    try core.changeMemberRole(groupId: gid, encryptedMemberId: emi, newRole: newRole)
                }.value
                await load()
            } catch {
                errorMessage = error.localizedDescription
            }
        }
    }

    /// Change the disappearing-message timer (admin-only; server-enforced).
    private func setExpiry(_ seconds: UInt32) {
        guard let core = appState.core(accountId: accountId) else { return }
        let gid = groupId
        Task {
            do {
                try await Task.detached {
                    try core.setGroupExpiry(groupId: gid, expirySeconds: seconds)
                }.value
                await load()
            } catch {
                errorMessage = error.localizedDescription
                await load()  // revert the picker to the server's value on failure
            }
        }
    }

    private func leaveGroup() {
        guard let core = appState.core(accountId: accountId) else { return }
        let gid = groupId
        Task {
            do {
                // Self-class leave (docs/53): works for any member, not just
                // admins. Tombstone-in-place — the group stays in the inbox,
                // read-only, with a "You left the group" entry as the last
                // message. Reload the timeline so that entry shows, then pop
                // back to the (now read-only) conversation.
                try await Task.detached {
                    try core.leaveGroup(groupId: gid)
                }.value
                appState.reloadGroupTimelineIfLoaded(groupId: gid, accountId: accountId)
                dismiss()
            } catch {
                errorMessage = error.localizedDescription
            }
        }
    }
}

#if DEBUG
#Preview {
    let me = Account(
        id: "did:plc:me",
        displayName: "Me",
        avatarData: nil,
        servers: [ServerInfo(
            id: "https://server.example",
            name: "Example",
            url: URL(string: "https://server.example")!
        )]
    )
    let contacts: [ContactRowFfi] = [
        ContactRowFfi(did: "did:plc:bob", displayName: "Bob Chena", isCurated: true, lastInteractionAtMs: 0),
        ContactRowFfi(did: "did:plc:carol", displayName: "Carol X", isCurated: true, lastInteractionAtMs: 0),
    ]
    let summary = GroupSummaryFfi(
        groupId: "grp1",
        masterKey: Data(count: 32),
        revision: 3,
        title: "March Logistics",
        description: "Planning crew for the day-of action.",
        expirySeconds: 0,
        // Self listed last on the wire — the view sorts "You" to the top.
        members: [
            GroupMemberFfi(did: "did:plc:bob", encryptedMemberId: "emi-bob", role: 0, joinedAtMs: 0),
            GroupMemberFfi(did: "did:plc:carol", encryptedMemberId: "emi-carol", role: 1, joinedAtMs: 0),
            GroupMemberFfi(did: "did:plc:me", encryptedMemberId: "emi-me", role: 1, joinedAtMs: 0),
        ],
        pendingInvites: [
            GroupPendingFfi(encryptedMemberId: "emi-dave", timestampMs: 0),
        ],
        pendingApprovals: []
    )
    return NavigationStack {
        GroupDetailView(groupId: "grp1", accountId: "did:plc:me")
            .environmentObject(AppState.preview(
                accounts: [me],
                contacts: contacts,
                groups: ["grp1": summary]
            ))
    }
}
#endif

File: mobile/android/app/src/main/kotlin/Views/Chats/GroupDetailView.kt
package net.theavalanche.app

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.ArrowBack
import androidx.compose.material.icons.filled.MoreVert
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.DropdownMenu
import androidx.compose.material3.DropdownMenuItem
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.material3.TopAppBar
import androidx.compose.material3.TopAppBarDefaults
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import uniffi.app_core.GroupMemberFfi
import uniffi.app_core.GroupPendingFfi
import uniffi.app_core.GroupSummaryFfi

// Group detail screen (docs/03-groups.md): member list, admin role changes,
// disappearing-message timer, and leave-group. Admin-only controls are shown
// only when the current user is an admin; the server enforces the same.
//
// Mirrors mobile/ios/Actnet/Sources/Views/Chats/GroupDetailView.swift.
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun GroupDetailView(
    groupId: String,
    accountId: String,
    appViewModel: AppViewModel,
    onDismiss: () -> Unit = {},
) {
    val scope = rememberCoroutineScope()

    var summary by remember { mutableStateOf<GroupSummaryFfi?>(null) }
    var loading by remember { mutableStateOf(true) }
    var errorMessage by remember { mutableStateOf<String?>(null) }
    // Separate var so we can guard against spurious initial onChange fires.
    var expirySeconds by remember { mutableStateOf(0u) }
    var showRename by remember { mutableStateOf(false) }
    var renameText by remember { mutableStateOf("") }

    // Computed from summary
    val amAdmin = summary?.members?.firstOrNull { it.did == accountId }?.role?.toInt() == 1
    val amMember = summary?.members?.any { it.did == accountId } ?: false

    // -----------------------------------------------------------------------
    // Helpers (mirrors private funcs in iOS)
    // -----------------------------------------------------------------------

    fun orderedMembers(members: List<GroupMemberFfi>): List<GroupMemberFfi> =
        members.sortedByDescending { it.did == accountId }

    fun memberName(member: GroupMemberFfi): String =
        if (member.did == accountId) "You"
        else appViewModel.resolvedName(did = member.did, accountId = accountId)

    fun isBot(member: GroupMemberFfi): Boolean =
        member.did != accountId && appViewModel.isBot(did = member.did, accountId = accountId)

    suspend fun load() {
        loading = true
        val core = appViewModel.core(accountId) ?: run {
            errorMessage = "No core for account"
            loading = false
            return
        }
        val gid = groupId
        try {
            val s = withContext(Dispatchers.IO) { core.fetchGroupState(groupId = gid) }
            summary = s
            expirySeconds = s.expirySeconds
        } catch (e: Exception) {
            // Server fetch is membership-gated; falls back to cached state so the
            // screen renders read-only after leaving (docs/53 §Leave).
            val cached = runCatching {
                withContext(Dispatchers.IO) { core.cachedGroupState(groupId = gid) }
            }.getOrNull()
            if (cached != null) {
                summary = cached
                expirySeconds = cached.expirySeconds
            } else {
                errorMessage = e.localizedMessage ?: e.message ?: "Unknown error"
            }
        }
        loading = false
    }

    fun setExpiry(seconds: UInt) {
        val core = appViewModel.core(accountId) ?: return
        val gid = groupId
        scope.launch {
            try {
                withContext(Dispatchers.IO) { core.setGroupExpiry(groupId = gid, expirySeconds = seconds) }
                load()
            } catch (e: Exception) {
                errorMessage = e.localizedMessage ?: e.message ?: "Unknown error"
                load()  // revert picker to server's value on failure
            }
        }
    }

    fun changeRole(member: GroupMemberFfi, toAdmin: Boolean) {
        val core = appViewModel.core(accountId) ?: return
        val gid = groupId
        val emi = member.encryptedMemberId
        val newRole: Short = if (toAdmin) 1 else 0
        scope.launch {
            try {
                withContext(Dispatchers.IO) {
                    core.changeMemberRole(groupId = gid, encryptedMemberId = emi, newRole = newRole)
                }
                load()
            } catch (e: Exception) {
                errorMessage = e.localizedMessage ?: e.message ?: "Unknown error"
            }
        }
    }

    fun renameGroup(name: String) {
        val trimmed = name.trim()
        val core = appViewModel.core(accountId) ?: return
        if (trimmed.isEmpty() || trimmed == (summary?.title ?: "")) return
        val gid = groupId
        scope.launch {
            try {
                withContext(Dispatchers.IO) { core.setGroupTitle(groupId = gid, newTitle = trimmed) }
                load()
            } catch (e: Exception) {
                errorMessage = e.localizedMessage ?: e.message ?: "Unknown error"
            }
        }
    }

    fun leaveGroup() {
        val core = appViewModel.core(accountId) ?: return
        val gid = groupId
        scope.launch {
            try {
                withContext(Dispatchers.IO) { core.leaveGroup(groupId = gid) }
                appViewModel.reloadGroupTimelineIfLoaded(groupId = gid, accountId = accountId)
                onDismiss()
            } catch (e: Exception) {
                errorMessage = e.localizedMessage ?: e.message ?: "Unknown error"
            }
        }
    }

    // Load on first composition
    LaunchedEffect(groupId, accountId) { load() }

    // -----------------------------------------------------------------------
    // Rename dialog — mirrors SwiftUI .alert("Rename group", ...)
    // -----------------------------------------------------------------------
    if (showRename) {
        AlertDialog(
            onDismissRequest = { showRename = false },
            title = { Text("Rename group") },
            text = {
                OutlinedTextField(
                    value = renameText,
                    onValueChange = { renameText = it },
                    label = { Text("Group name") },
                    singleLine = true,
                )
            },
            confirmButton = {
                TextButton(
                    onClick = {
                        renameGroup(renameText)
                        showRename = false
                    },
                    enabled = renameText.trim().isNotEmpty(),
                ) {
                    Text("Save")
                }
            },
            dismissButton = {
                TextButton(onClick = { showRename = false }) { Text("Cancel") }
            },
        )
    }

    // -----------------------------------------------------------------------
    // Error dialog for action failures
    // -----------------------------------------------------------------------
    val errMsg = errorMessage
    if (errMsg != null) {
        AlertDialog(
            onDismissRequest = { errorMessage = null },
            title = { Text("Error") },
            text = { Text(errMsg) },
            confirmButton = {
                TextButton(onClick = { errorMessage = null }) { Text("OK") }
            },
        )
    }

    // -----------------------------------------------------------------------
    // Screen scaffold
    // -----------------------------------------------------------------------
    Scaffold(
        topBar = {
            TopAppBar(
                title = { Text("Group info") },
                navigationIcon = {
                    IconButton(onClick = onDismiss) {
                        Icon(Icons.AutoMirrored.Filled.ArrowBack, contentDescription = "Back")
                    }
                },
                colors = TopAppBarDefaults.topAppBarColors(
                    containerColor = LocalAvalancheColors.current.paper,
                    titleContentColor = LocalAvalancheColors.current.ink,
                ),
            )
        },
        containerColor = LocalAvalancheColors.current.paper,
    ) { innerPadding ->
        val s = summary
        when {
            s != null -> {
                LazyColumn(
                    modifier = Modifier
                        .fillMaxSize()
                        .background(LocalAvalancheColors.current.paper)
                        .padding(innerPadding),
                ) {
                    // -------------------------------------------------------
                    // Section 1: title / description / revision
                    // -------------------------------------------------------
                    item {
                        SectionContainer {
                            Column(
                                modifier = Modifier
                                    .fillMaxWidth()
                                    .padding(horizontal = 16.dp, vertical = 12.dp),
                                verticalArrangement = Arrangement.spacedBy(4.dp),
                            ) {
                                Row(
                                    modifier = Modifier.fillMaxWidth(),
                                    verticalAlignment = Alignment.CenterVertically,
                                ) {
                                    Text(
                                        text = if (s.title.isEmpty()) "Group" else s.title,
                                        style = MaterialTheme.typography.titleMedium,
                                        modifier = Modifier.weight(1f),
                                    )
                                    if (amAdmin) {
                                        TextButton(onClick = {
                                            renameText = s.title
                                            showRename = true
                                        }) {
                                            Text(
                                                text = "Rename",
                                                fontSize = 14.sp,
                                                color = LocalAvalancheColors.current.brand,
                                            )
                                        }
                                    }
                                }
                                if (s.description.isNotEmpty()) {
                                    Text(
                                        text = s.description,
                                        style = MaterialTheme.typography.bodyMedium,
                                        color = LocalAvalancheColors.current.muted,
                                    )
                                }
                                Text(
                                    text = "Revision ${s.revision}",
                                    style = MaterialTheme.typography.labelSmall,
                                    color = LocalAvalancheColors.current.muted,
                                )
                            }
                        }
                        SectionDivider()
                    }

                    // -------------------------------------------------------
                    // Section 2: disappearing messages
                    // -------------------------------------------------------
                    item {
                        SectionHeader("Disappearing messages")
                        SectionContainer {
                            if (amAdmin) {
                                val loadedExpiry = s.expirySeconds
                                DisappearingMessagesPickerView(
                                    seconds = expirySeconds,
                                    onSecondsChange = { newValue ->
                                        // Guard: ignore the initial seeding; only react on a real change.
                                        if (newValue != loadedExpiry) {
                                            expirySeconds = newValue
                                            setExpiry(newValue)
                                        } else {
                                            expirySeconds = newValue
                                        }
                                    },
                                    modifier = Modifier
                                        .fillMaxWidth()
                                        .padding(horizontal = 16.dp, vertical = 8.dp),
                                )
                            } else {
                                Row(
                                    modifier = Modifier
                                        .fillMaxWidth()
                                        .padding(horizontal = 16.dp, vertical = 12.dp),
                                    verticalAlignment = Alignment.CenterVertically,
                                ) {
                                    Text(
                                        text = "Timer",
                                        style = MaterialTheme.typography.bodyMedium,
                                        modifier = Modifier.weight(1f),
                                    )
                                    Text(
                                        text = DisappearingMessagesPicker.label(s.expirySeconds),
                                        style = MaterialTheme.typography.bodyMedium,
                                        color = LocalAvalancheColors.current.muted,
                                    )
                                }
                            }
                        }
                        SectionDivider()
                    }

                    // -------------------------------------------------------
                    // Section 3: members
                    // -------------------------------------------------------
                    item {
                        SectionHeader("Members (${s.members.size})")
                    }
                    items(orderedMembers(s.members), key = { it.encryptedMemberId }) { member ->
                        SectionContainer {
                            MemberRow(
                                name = memberName(member),
                                isBot = isBot(member),
                                role = member.role,
                                isSelf = member.did == accountId,
                                amAdmin = amAdmin,
                                onMakeAdmin = { changeRole(member, toAdmin = true) },
                                onRemoveAdmin = { changeRole(member, toAdmin = false) },
                            )
                        }
                        HorizontalDivider(
                            modifier = Modifier.padding(start = 58.dp),
                            color = LocalAvalancheColors.current.divider,
                        )
                    }

                    // -------------------------------------------------------
                    // Section 4: pending invites (if any)
                    // -------------------------------------------------------
                    if (s.pendingInvites.isNotEmpty()) {
                        item {
                            SectionDivider()
                            SectionHeader("Pending invites (${s.pendingInvites.size})")
                        }
                        items(s.pendingInvites, key = { it.encryptedMemberId }) { pending ->
                            SectionContainer {
                                Text(
                                    text = pending.encryptedMemberId,
                                    style = MaterialTheme.typography.labelSmall,
                                    color = LocalAvalancheColors.current.muted,
                                    maxLines = 1,
                                    modifier = Modifier
                                        .fillMaxWidth()
                                        .padding(horizontal = 16.dp, vertical = 12.dp),
                                )
                            }
                        }
                    }

                    // -------------------------------------------------------
                    // Section 5: leave / "you left" notice
                    // -------------------------------------------------------
                    item {
                        SectionDivider()
                        SectionContainer {
                            if (amMember) {
                                TextButton(
                                    onClick = { leaveGroup() },
                                    colors = ButtonDefaults.textButtonColors(
                                        contentColor = LocalAvalancheColors.current.error,
                                    ),
                                    modifier = Modifier
                                        .fillMaxWidth()
                                        .padding(horizontal = 8.dp, vertical = 4.dp),
                                ) {
                                    Text("Leave group")
                                }
                            } else {
                                Text(
                                    text = "You left this group.",
                                    style = MaterialTheme.typography.labelSmall,
                                    color = LocalAvalancheColors.current.muted,
                                    modifier = Modifier
                                        .fillMaxWidth()
                                        .padding(horizontal = 16.dp, vertical = 12.dp),
                                )
                            }
                        }
                    }
                }
            }

            loading -> {
                Box(
                    modifier = Modifier
                        .fillMaxSize()
                        .padding(innerPadding),
                    contentAlignment = Alignment.Center,
                ) {
                    CircularProgressIndicator(color = LocalAvalancheColors.current.brand)
                }
            }

            else -> {
                // Error / unavailable state
                Box(
                    modifier = Modifier
                        .fillMaxSize()
                        .padding(innerPadding)
                        .padding(24.dp),
                    contentAlignment = Alignment.Center,
                ) {
                    Column(
                        horizontalAlignment = Alignment.CenterHorizontally,
                        verticalArrangement = Arrangement.spacedBy(8.dp),
                    ) {
                        Text(
                            text = "Couldn't load group",
                            style = MaterialTheme.typography.titleMedium,
                            fontWeight = FontWeight.SemiBold,
                        )
                        if (errMsg != null) {
                            Text(
                                text = errMsg,
                                style = MaterialTheme.typography.bodyMedium,
                                color = LocalAvalancheColors.current.muted,
                            )
                        }
                    }
                }
            }
        }
    }
}

// ---------------------------------------------------------------------------
// Private composable helpers
// ---------------------------------------------------------------------------

@Composable
private fun SectionHeader(title: String) {
    Text(
        text = title.uppercase(),
        style = MaterialTheme.typography.labelSmall,
        color = LocalAvalancheColors.current.muted,
        modifier = Modifier.padding(horizontal = 16.dp, vertical = 8.dp),
    )
}

@Composable
private fun SectionContainer(content: @Composable () -> Unit) {
    Box(
        modifier = Modifier
            .fillMaxWidth()
            .background(LocalAvalancheColors.current.card),
    ) {
        content()
    }
}

@Composable
private fun SectionDivider() {
    HorizontalDivider(
        color = LocalAvalancheColors.current.divider,
        thickness = 0.5.dp,
        modifier = Modifier.padding(vertical = 8.dp),
    )
}

@Composable
private fun MemberRow(
    name: String,
    isBot: Boolean,
    role: Short,
    isSelf: Boolean,
    amAdmin: Boolean,
    onMakeAdmin: () -> Unit,
    onRemoveAdmin: () -> Unit,
) {
    var menuExpanded by remember { mutableStateOf(false) }

    Row(
        modifier = Modifier
            .fillMaxWidth()
            .padding(horizontal = 16.dp, vertical = 10.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(10.dp),
    ) {
        ContactAvatar(name = name, isBot = isBot, size = 32.dp)
        Text(
            text = name,
            style = MaterialTheme.typography.bodyMedium,
            modifier = Modifier.weight(1f),
            maxLines = 1,
        )
        // Admin badge
        if (role == 1.toShort()) {
            Text(
                text = "Admin",
                style = MaterialTheme.typography.labelSmall,
                color = LocalAvalancheColors.current.brand,
                modifier = Modifier
                    .clip(RoundedCornerShape(50))
                    .background(LocalAvalancheColors.current.brand.copy(alpha = 0.15f))
                    .padding(horizontal = 6.dp, vertical = 2.dp),
            )
        }
        // Per-member admin menu — visible only to admins and not for self
        if (amAdmin && !isSelf) {
            Box {
                IconButton(onClick = { menuExpanded = true }) {
                    Icon(
                        imageVector = Icons.Default.MoreVert,
                        contentDescription = "Member options",
                        tint = LocalAvalancheColors.current.muted,
                    )
                }
                DropdownMenu(
                    expanded = menuExpanded,
                    onDismissRequest = { menuExpanded = false },
                ) {
                    if (role == 1.toShort()) {
                        DropdownMenuItem(
                            text = { Text("Remove admin") },
                            onClick = {
                                menuExpanded = false
                                onRemoveAdmin()
                            },
                        )
                    } else {
                        DropdownMenuItem(
                            text = { Text("Make admin") },
                            onClick = {
                                menuExpanded = false
                                onMakeAdmin()
                            },
                        )
                    }
                }
            }
        }
    }
}

// ---------------------------------------------------------------------------
// Preview
// ---------------------------------------------------------------------------

@Preview(showBackground = true)
@Composable
private fun GroupDetailViewPreview() {
    AvalancheTheme {
        // The preview ViewModel's mock core returns no group summary, so this
        // renders the loading state — useful for reviewing the scaffold/top bar.
        GroupDetailView(
            groupId = "preview-group",
            accountId = "did:example:alice",
            appViewModel = rememberPreviewAppViewModel(),
        )
    }
}
